src/network/client.rs

```diff
@@ -230,13 +349,82 @@ impl NetworkClient {
+        let Some(assignment) = local_assignments.get(&heartbeat.assignment_id) else {
+            tracing::debug!("Assignment {:?} not found", heartbeat.assignment_id);
+            let latest_assignmaent_id = local_assignments
+                .last_key_value()
+                .map(|(assignment_id, _)| assignment_id.clone())
+                .unwrap_or_default();
+            if heartbeat.assignment_id > latest_assignmaent_id {
+                self.heartbeat_buffer.lock().push_back((peer_id, heartbeat));
+                tracing::debug!("Putting heartbeat into waitlist for {peer_id}");
+                {
+                    let mut heartbeat_buffer = self.heartbeat_buffer.lock();
+                    if heartbeat_buffer.len() > MAX_WAITING_PINGS {
+                        heartbeat_buffer.pop_front();
+                    }
+                }
+            } else {
+                tracing::error!("Dropping heartbeat from {peer_id}");
+            }
+            return;
+        };
+        let Some(chunk_list) = assignment.get(&peer_id.to_string()) else {
+            tracing::error!(
+                "PeerID {:?} not found in Assignment {:?}",
+                peer_id,
+                heartbeat.assignment_id
+            );
+            return;
+        };
+        let Some(missing_chunks) = heartbeat.missing_chunks.as_ref() else {
+            tracing::error!(
+                "PeerID {:?}: missing_chunks are missing in heartbeat",
+                peer_id
+            );
+            return;
+        };
+        let unavailability_map = missing_chunks.to_bytes();
+        if unavailability_map.len() != chunk_list.len() {
+            tracing::error!(
+                "Heartbeat of {:?} and assignment {:?} are inconsistent",
+                peer_id,
+                heartbeat.assignment_id
+            );
             return;
         }
-        tracing::trace!("Ping from {peer_id}");
+        let worker_state = zip(unavailability_map, chunk_list)
+            .filter_map(|(is_missing, val)| {
+                if is_missing == 0 {
+                    Some(val.clone())
+                } else {
+                    None
+                }
+            })
+            .group_by(|(dataset, _)| dataset.clone())
+            .into_iter()
+            .map(|(group, vals)| {
+                (
+                    group,
+                    RangeSet {
+                        ranges: vals.map(|(_, range)| range).sorted().collect(),
+                    },
+                )
+            })
+            .collect::<HashMap<_, _>>();
+        self.network_state
+            .lock()
+            .update_dataset_states(peer_id, worker_state);
         metrics::VALID_PINGS.inc();
-        todo!();
+        //tracing::error!("Payload: {:?}", heartbeat.missing_chunks.as_ref().unwrap().to_bytes().iter().map(|v| *v as usize).sum::<usize>());
+        // if !heartbeat.version_matches(&SUPPORTED_VERSIONS) {
```

Comment:
You forgot to check the version it seems

Reply:
done